### DISCUSSION TITLE
How can I identify styles being used?

### DISCUSSION THREAD

**Discussion by saiwing-yeung** [Textualize/rich#1564] [Q&A] [answered]
Is there a way to identify the style being used? I noticed that time strings are in bright green and for me it's quite hard to read.

![image](https://user-images.githubusercontent.com/12464429/136827185-2b6fab00-e29b-4b4e-b782-d2de516c5569.png)

I checked `python -m rich.theme` but nothing seem obvious (there is one style with the name `log.time` but that's not it). How do I figure out which style I should change? Thanks!

**Comment by saschalalala** [comment, marked as answer]
## Spoiler

Found the solution while I wrote this answer. It is `repr.ipv6`, try the following:

```python
from rich.theme import Theme
from rich.console import Console

console = Console(theme=Theme({"repr.ipv6": "red italic on yellow"}))

console.print("12:34:45")
```

## My digging around before finding the solution

After digging around in the source code and learning about Style and Theme and Color classes, it seems to me as if this string is some kind of special case:

<img width="419" alt="grafik" src="https://user-images.githubusercontent.com/3439311/137013537-cf069a71-09a4-4dbc-804b-7532ea7d8195.png">

At least I found out that the used color is green (not cyan, which is the color for log.time).

I found out that the only colors that a string like this reacts to is the background color:

<img width="482" alt="grafik" src="https://user-images.githubusercontent.com/3439311/137013879-7de62a36-7448-452d-9544-1f5cfd0bbcb3.png">

which, for my understanding is caused by syntax highlighting, see the following example:

`console.print("This is a test 12:34:56 [1,2,3]", style="bold red")`

<img width="579" alt="grafik" src="https://user-images.githubusercontent.com/3439311/137013932-0fedbc57-1cad-4fcd-a22c-a9e840e9237b.png">

What _does_ work, anyways, is the following:

`console.print("[bold red]12:34:56[/bold red]")` which prints the timestamp in bold red

**Reply by saiwing-yeung** [reply]
Thanks so much! 12:34:56 doesn't seem to be a valid IPv6 address. So maybe we should adjust the regex in `highlighter.py`.

And I guess there is still no general solution to this — the next time someone runs into a situation where an unknown syntax is triggered there is no easy way to identify which syntax it is.

**Reply by willmcgugan** [reply]
The time there is being styles as an ipv6 address, but you're right that the regex is over simplified. A perfect ipv6 regex is absurdly complicated, but maybe there is a compromise.

The problem might not be exclusively within Rich though. Terminals usually have well chosen color themes where green on default is quite readable. You might want to tweak your color theme as green on a default background is a common combination.

**Reply by saschalalala** [reply]
> Thanks so much! 12:34:56 doesn't seem to be a valid IPv6 address. So maybe we should adjust the regex in `highlighter.py`.
> 
> And I guess there is still no general solution to this — the next time someone runs into a situation where an unknown syntax is triggered there is no easy way to identify which syntax it is.

I attached a debugger and found out that every renderable seems to have a list of spans which hold the information about the currently used rendering theme. If you replace the print method in rich's `console.py` with the following gist: https://gist.github.com/saschalalala/994273a288f1136a2888c8676b0873e9, you can do the following:

```python
from rich.console import Console

console = Console()
console.print('This is a test: 12:34:56, [1,2,3], {"asdf": "test"}', debug_style=True)
```

which generates the following output:

```
Span(26, 27, 'repr.brace')
Span(32, 33, 'repr.brace')
Span(35, 36, 'repr.brace')
Span(50, 51, 'repr.brace')
Span(16, 24, 'repr.ipv6')
Span(27, 28, 'repr.number')
Span(29, 30, 'repr.number')
Span(31, 32, 'repr.number')
Span(36, 42, 'repr.str')
Span(44, 50, 'repr.str')
This is a test: 12:34:56, [1,2,3], {"asdf": "test"}
```

This isn't 100% practical yet but could be worth a PR if implemented a little nicer.

**Reply by saschalalala** [reply]
You can also debug this yourself by setting a breakpoint here: https://github.com/willmcgugan/rich/blob/master/rich/console.py#L1577 and inspect the renderables on your own if you want to find out more (Screenshot from VS Code):

<img width="609" alt="grafik" src="https://user-images.githubusercontent.com/3439311/137123607-df9ab44a-9508-4c0c-99e8-e4e17214be05.png">

**Reply by willmcgugan** [reply]
Good sleuthing! 

That has jogged my memory. Here's an easier way to get that same information:

```python
from rich.console import Console
console = Console()
text = console.render_str('This is a test: 12:34:56, [1,2,3], {"asdf": "test"}')
console.print(text.spans)
```

**Reply by saschalalala** [reply]
That's much easier :D

**Reply by saiwing-yeung** [reply]
Awesome. Thanks to you both!
